### ISSUE TITLE
Add regexp to MagicNumber rule

### ISSUE THREAD

**Issue by apachucy** [detekt/detekt#4080]
Right now we can use MagicNumber rule with defined ignoreNumbers that are intigers. Is there a chance to add regexp to ignoreNumbers?

I am interested in case when you need to pass hex (for example to colors).
This case will trigger MagicNumber rule: Color( 0x66000000)
This case will not trigger MagicNumber rule: Color(color = 0x66000000) but I prefer using the previous example 

Cheers,
AP

**Comment by cortinico** [comment]
> Is there a chance to add regexp to ignoreNumbers?

I understand your use case, though the rule has already a lot of ignores.
Plus your color example looks like the classical example where you want to export the number to a constant:

```
val COLOR_ALPHA_40 = 0x66000000
```

**Comment by schalkms** [comment]
I agree with Nico. I don’t think we should add code to this rule for this special case. This should be an @Ignore or a custom rule by copying the detekt one.

**Comment by BraisGabin** [comment]
To me this use case should be "allow magic numbers as a parameter of `Color.constructor`". That could be archived if we implement #4068.

**Comment by ccjernigan** [comment]
Detekt's magic number check is also painful with the Kotlin Duration API.  For example, the following declaration will get flagged as a magic number: `val timeout = Duration.seconds(10)`.  At least in my particular codebase, we've currently got over 100 `@Suppress` annotations on durations alone.

I realize the Duration API is not yet stable, although it should be stabilized with the next Kotlin release.

The ability to tune the MagicNumber rule, either with a regex or just a hard-coded exclusion for the Duration API, would make this much less painful to work with.

Would the Detekt maintainers prefer that I open a new issue specifically for the Duration API?  Or does this use case add weight to the regex solution proposed by this issue?

**Comment by marschwar** [comment]
To me this is the same issue and does not anything. The idea behind the MagicNumber rule is to improve readability. In your particular example this should be either `TEN_SECONDS` or `DEFAULT_TIMEOUT`. Both carry a (different) meaning.

**Comment by BraisGabin** [comment]
Does `Duration.seconds(TEN_SECONDS)` improve the readability over `Duration.seconds(10)`? For things like this I have this rule disabled. But I insist that #4068 should fix that. We should find which type of ignores we need.

**Comment by marschwar** [comment]
> Does Duration.seconds(TEN_SECONDS) improve the readability over Duration.seconds(10)?

No, that would be horrible.

I was talking about `val TEN_SECONDS = Duration.seconds(10)`. Does this fail as well? If so, I will change my opinion.

**Comment by chao2zhang** [comment]
What I am comparing is the checkstyle's `MagicNumber.ignoreFieldDeclartion`
https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.html

I believe we should be able to add a configurable option to allow magic number for class/companion object/top-level properties

**Comment by BraisGabin** [comment]
I think that all this cases could be solved with something similar to #4148. That suppressor works on the function declaration. In our case we would need to create one for when a fuction is called. So all it's paramaters are ignored (including code inside lambdas). I think that all the problems exposed in this issue would be solved with that.

It should look like:

```yaml
ignoreFunctionCall:
  - 'android.compose.Color.constructor(Int)'
  - 'kotlinx.time.Duration.seconds(Long)' 
```

(I just guessed the full qualified name of those functions)

And, of course, it will work for all the rules in detekt. Not just for `MagicNumber`. What do you think?

**Comment by chao2zhang** [comment]
@BraisGabin I think the `ignoreFunctionCall` would work generally. However, for this particular `MagicNumber` rule, what it makes sense is to ignore when the magic number is used to initialize a class / top-level property.

**Comment by BraisGabin** [comment]
Initialize a class can be handled too with `constructor`. And the top level property, when there is not call, is already handled by the rule. That's the only case that this rule allows.

**Comment by chao2zhang** [comment]
I think there are two problems here:

1. Leveraging `ignorePropertyDeclaration`: I just added https://github.com/detekt/detekt/pull/4152 to verify that they are working as expected.
```
class A {
  val duration = Duration.seconds(10) // NO VIOLATION if ignorePropertyDeclaration = true

  companion object {
    val color = Color(0x66000000) // NO VIOLATION if ignoreCompanionObjectPropertyDeclaration = true by default
  }
}
```
This is what being discussed by @ccjernigan and @marschwar 

2. Ignoring specific function calls as @BraisGabin was mentioning.

For the original author @apachucy, please let us know if solution 1 mentioned here works for you.

**Comment by BraisGabin** [comment]
The part mentioned by me was "theoretical". So, if the other points are already covered by this rule we should close this, right?
